In [3]:
import pandas as pd
print(pd.__version__)

3.0.5


In [4]:
import os 
print(os.getcwd())

e:\Coading\job-posting-authenticity-detector\notebooks


In [5]:
emscad = pd.read_csv("../data/emscad_core.csv")
synth = pd.read_csv("../data/synthetic_stress_test.csv")

In [6]:
print("EMSCAD columns:", emscad.columns.tolist())
print("\nSynthetic columns:", synth.columns.tolist())

EMSCAD columns: ['job_id', 'title', 'location', 'department', 'salary_range', 'company_profile', 'description', 'requirements', 'benefits', 'telecommuting', 'has_company_logo', 'has_questions', 'employment_type', 'required_experience', 'required_education', 'industry', 'function', 'fraudulent']

Synthetic columns: ['job_id', 'job_title', 'job_description', 'requirements', 'benefits', 'company_name', 'company_profile', 'industry', 'employment_type', 'location', 'salary_range', 'required_experience_years', 'education_level', 'department', 'posting_date', 'application_deadline', 'contact_email', 'company_website', 'has_logo', 'num_open_positions', 'job_function', 'telecommuting', 'fraud_reason', 'text_length', 'is_fake']


In [7]:
rename_map = {
    "job_title": "title",
    "job_description" : "description",
    "has_logo" : "has_company_logo",
    "job_function": "function",
    "is_fake" : "fraudulent",
    "education_level" : "required_education"
}

In [8]:
synth_renamed = synth.rename(columns=rename_map)

In [9]:
emscad["source"] =  "emscad"
synth_renamed["source"] = "synthetic"

In [10]:
# This shows us exact categorical lables EMSCAD uses
print(emscad["required_experience"].value_counts(dropna=False))

required_experience
NaN                 7050
Mid-Senior level    3809
Entry level         2697
Associate           2297
Not Applicable      1116
Director             389
Internship           381
Executive            141
Name: count, dtype: int64


In [11]:
print(synth["required_experience_years"].describe())

count    3000.000000
mean        5.044000
std         3.168292
min         0.000000
25%         2.000000
50%         5.000000
75%         8.000000
max        10.000000
Name: required_experience_years, dtype: float64


## The Bucketing of data (years of experience-> job roles)

since the emscad required ewperience goes to
required_experience
NaN                 7050
Mid-Senior level    3809
Entry level         2697
Associate           2297
Not Applicable      1116
Director             389
Internship           381
Executive            141

but require experience years in synthetic data only goes to max 10 years , that means directorial roles are not achievable , so we bucket roles to experince insted of dividing the roles to fit into experience brackets . this will be more appropriate for the data analysis but some rows of synthetic wil remain unmapped.

Proposed boundaries
Years	EMSCAD label
0–1	Entry level
2–4	Associate
5–10	Mid-Senior level

In [12]:
#Bucket Function
def bucket_experience(years):
    if years <=1:
        return "Entry level"
    elif years <=4:
        return "Associate"
    else :
        return "Mid-Senior level"
    
synth_renamed["required_experience"] = synth_renamed["required_experience_years"].apply(bucket_experience)
# print(synth_renamed["required_experience"].value_counts())
print(synth_renamed.columns.tolist())

['job_id', 'title', 'description', 'requirements', 'benefits', 'company_name', 'company_profile', 'industry', 'employment_type', 'location', 'salary_range', 'required_experience_years', 'required_education', 'department', 'posting_date', 'application_deadline', 'contact_email', 'company_website', 'has_company_logo', 'num_open_positions', 'function', 'telecommuting', 'fraud_reason', 'text_length', 'fraudulent', 'source', 'required_experience']


In [13]:
#Empty set confirms it — every label in your bucketed synthetic column now exists exactly in EMSCAD's vocabulary
print(set(synth_renamed["required_experience"].unique())-set(emscad["required_experience"].unique()))

set()


In [14]:
emscad["source"] =  "emscad"
synth_renamed["source"] =  "synthetic"

In [15]:
print(emscad.info())
print(synth_renamed.info())

<class 'pandas.DataFrame'>
RangeIndex: 17880 entries, 0 to 17879
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   job_id               17880 non-null  int64
 1   title                17880 non-null  str  
 2   location             17534 non-null  str  
 3   department           6333 non-null   str  
 4   salary_range         2868 non-null   str  
 5   company_profile      14572 non-null  str  
 6   description          17879 non-null  str  
 7   requirements         15184 non-null  str  
 8   benefits             10668 non-null  str  
 9   telecommuting        17880 non-null  int64
 10  has_company_logo     17880 non-null  int64
 11  has_questions        17880 non-null  int64
 12  employment_type      14409 non-null  str  
 13  required_experience  10830 non-null  str  
 14  required_education   9775 non-null   str  
 15  industry             12977 non-null  str  
 16  function             11425 non-nu

In [16]:
print(synth_renamed.columns.tolist())

['job_id', 'title', 'description', 'requirements', 'benefits', 'company_name', 'company_profile', 'industry', 'employment_type', 'location', 'salary_range', 'required_experience_years', 'required_education', 'department', 'posting_date', 'application_deadline', 'contact_email', 'company_website', 'has_company_logo', 'num_open_positions', 'function', 'telecommuting', 'fraud_reason', 'text_length', 'fraudulent', 'source', 'required_experience']


In [17]:
synth_renamed = synth.rename(columns=rename_map)
synth_renamed["source"] = "synthetic"
print(synth_renamed.columns.tolist())

['job_id', 'title', 'description', 'requirements', 'benefits', 'company_name', 'company_profile', 'industry', 'employment_type', 'location', 'salary_range', 'required_experience_years', 'required_education', 'department', 'posting_date', 'application_deadline', 'contact_email', 'company_website', 'has_company_logo', 'num_open_positions', 'function', 'telecommuting', 'fraud_reason', 'text_length', 'fraudulent', 'source']


# messiness of dataset
look at how differently missing values are distributed between the two datasets. EMSCAD has real gaps everywhere — salary_range only 2,868/17,880 filled, benefits 10,668/17,880, department 6,333/17,880. The synthetic dataset has almost no missing values except company_name, company_website, and fraud_reason. That's not a coincidence — it's a direct fingerprint of the two datasets' origins: EMSCAD is messy because real recruiters fill out forms inconsistently; the synthetic data is clean because whatever generated it didn't model that real-world sloppiness. That's actually a good observed insight to write down now — it strengthens your earlier point about treating synthetic data as a separate stress-test set rather than blending it in as if it were equally realistic.

In [18]:
emscad["salary_missing"] = emscad["salary_range"].isnull()
print(emscad.groupby("salary_missing")["fraudulent"].mean())

salary_missing
False    0.077755
True     0.042832
Name: fraudulent, dtype: float64


# Strange finding
Read it carefully: salary_missing = False means salary is present, and that group has the higher fraud rate (7.78%) compared to salary_missing = True (missing salary, only 4.28%). That's the opposite of what "missing salary = vague/suspicious" would predict.

Think about why, using the same reasoning we used for has_company_logo: scammers want to lure you in with something concrete and appealing — an unrealistically high salary is classic bait ("$5000/month, work from home, no experience needed"). Legitimate companies, especially ones being careful and professional, often deliberately withhold salary until later stages, exactly like your hypothesis said — but that same caution is actually a sign of legitimacy, not vagueness. Meanwhile, fraudulent postings tend to flaunt salary because that's the whole hook.

In [19]:
# This prints 10 examples of salary that are not null.
print(emscad[emscad["salary_missing"]== False]["salary_range"].head(10))

6       20000-28000
10    100000-120000
15    120000-150000
23    100000-120000
31      50000-65000
42      40000-50000
65            60-80
76      65000-70000
77           75-115
79     75000-110000
Name: salary_range, dtype: str


# observation on salary
the smaller ranges like 60 -80 , 75-115 are not broken but hourly rates.

In [20]:
bad_rows = salaries[salaries.str.contains("[A-Za-z]", na=False)]
print(bad_rows.head(20))
print(f"\nTotal bad rows: {len(bad_rows)}")

NameError: name 'salaries' is not defined

In [ ]:
# fix for Excel date-corruption pattern
clean_salaries = salaries[~salaries.str.contains("[A-Za-z]", na=False)]
lower_bound = clean_salaries.str.split("-").str[0].astype(float)
print(lower_bound.describe())

count    2.842000e+03
mean     5.154100e+05
std      1.769313e+07
min      0.000000e+00
25%      1.800000e+04
50%      3.500000e+04
75%      6.000000e+04
max      8.000000e+08
Name: salary_range, dtype: float64


In [ ]:
print(lower_bound.sort_values().head(80).tail(50))

12058    0.0
5734     0.0
810      0.0
8910     0.0
12417    0.0
12419    0.0
12424    0.0
3475     0.0
11451    0.0
11379    0.0
3564     0.0
707      0.0
4293     0.0
3561     0.0
3642     0.0
12284    0.0
12088    0.0
12119    0.0
12434    0.0
7354     0.0
763      0.0
767      0.0
12501    0.0
12790    0.0
12849    0.0
5821     0.0
11255    0.0
4306     0.0
11229    0.0
11203    0.0
12389    0.0
16851    0.0
3442     0.0
12596    0.0
7837     0.0
13523    0.0
2917     0.0
4375     0.0
10997    0.0
11068    0.0
10885    0.0
10887    0.0
10902    0.0
10911    0.0
10918    0.0
5797     0.0
12857    0.0
12704    0.0
12947    0.0
5904     0.0
Name: salary_range, dtype: float64


In [ ]:
zero_count = (lower_bound == 0).sum()
print(f"Total zero values: {zero_count}")

print(lower_bound[lower_bound > 0].sort_values().head(30))

Total zero values: 195
14769     7.0
17538    13.0
3345     13.0
4043     13.0
17722    13.0
8162     13.0
1662     13.0
17195    13.0
8476     14.0
17698    15.0
6228     15.0
17700    15.0
16765    15.0
17783    15.0
17389    16.0
9332     17.0
7818     17.0
17775    17.0
17755    17.0
3342     18.0
7160     20.0
17638    20.0
14063    20.0
17666    20.0
3786     20.0
2930     20.0
6122     20.0
1843     21.0
4565     21.0
17724    21.0
Name: salary_range, dtype: float64


In [ ]:
lower_bound_clean = lower_bound[lower_bound >= 1000]
print(lower_bound_clean.describe())
print(f"\nRows excluded as likely non-annual: {(lower_bound < 1000).sum()}")

count    2.446000e+03
mean     5.988461e+05
std      1.907089e+07
min      1.000000e+03
25%      2.500000e+04
50%      4.000000e+04
75%      6.500000e+04
max      8.000000e+08
Name: salary_range, dtype: float64

Rows excluded as likely non-annual: 396


**Salary data cleaning decision:** Excluded 26 rows with Excel date-corrupted values 
(e.g. "Oct-15" instead of "10-15"). Excluded 396 additional rows with lower-bound 
salary under 1000 as likely non-annual pay units (hourly/weekly); exact unit could 
not be reliably determined from the data alone. Remaining outliers (e.g. max of 
800M) still need investigation.

In [ ]:
import sys
sys.path.append("../src")
from preprocessing import load_and_clean

df = load_and_clean("../data/emscad_core.csv")

candidates = ["has_company_logo", "has_questions", "telecommuting",
              "employment_type", "required_experience",
              "required_education", "salary_missing"]

for col in candidates:
    summary = df.groupby(col, dropna=False)["fraudulent"].agg(["mean", "count"]).round(3)
    print(f"\n--- {col} ---")
    print(summary)


--- has_company_logo ---
                   mean  count
has_company_logo              
0                 0.137   3088
1                 0.018  11714

--- has_questions ---
                mean  count
has_questions              
0              0.058   7494
1              0.028   7308

--- telecommuting ---
                mean  count
telecommuting              
0              0.042  14176
1              0.053    626

--- employment_type ---
                  mean  count
employment_type              
Contract         0.041    783
Full-time        0.039  10073
Other            0.044    206
Part-time        0.098    692
Temporary        0.010    193
NaN              0.045   2855

--- required_experience ---
                      mean  count
required_experience              
Associate            0.012   2054
Director             0.036    361
Entry level          0.060   2128
Executive            0.056    125
Internship           0.029    349
Mid-Senior level     0.028   3435
Not Applicable

In [22]:
import preprocessing
import importlib
importlib.reload(preprocessing)
from preprocessing import load_and_clean, combine_text_fields, build_training_features, STRUCTURED_COLUMNS

In [23]:
from preprocessing import load_and_clean, combine_text_fields, build_training_features, STRUCTURED_COLUMNS

df = combine_text_fields(load_and_clean("../data/emscad_core.csv"))
X = build_training_features(df)

print(X.shape)
print(X.isna().sum().sum())
for col in STRUCTURED_COLUMNS:
    print(f"\n{col}\n{X[col].value_counts(dropna=False)}")

(14802, 7)
0

has_company_logo
has_company_logo
yes    11714
no      3088
Name: count, dtype: int64

has_screening_questions
has_screening_questions
no     7494
yes    7308
Name: count, dtype: int64

salary_listed
salary_listed
no     12245
yes     2557
Name: count, dtype: int64

employment_type
employment_type
full_time    10073
unknown       2855
contract       783
part_time      692
other          206
temporary      193
Name: count, dtype: int64

required_education
required_education
unknown          8124
bachelors        4030
high_school      1616
other             490
masters           399
certification     143
Name: count, dtype: int64

required_experience
required_experience
unknown               5416
mid_senior            3435
entry                 2128
associate             2054
not_applicable         934
director_executive     486
internship             349
Name: count, dtype: int64


In [25]:
import joblib
pipeline = joblib.load("../models/fraud_pipeline.joblib")
feature_names = pipeline.named_steps["features"].get_feature_names_out()
coefs = pipeline.named_steps["clf"].coef_[0]

import pandas as pd
coef_df = pd.DataFrame({"feature": feature_names, "coefficient": coefs})
print(coef_df[coef_df["feature"].str.contains("cat__")].sort_values("coefficient", ascending=False))

                                           feature  coefficient
20000                     cat__has_company_logo_no     1.002233
20014          cat__required_education_high_school     0.719662
20008                   cat__employment_type_other     0.520875
20024             cat__required_experience_unknown     0.445523
20006                cat__employment_type_contract     0.434182
20013        cat__required_education_certification     0.428128
20021          cat__required_experience_internship     0.399321
20005                       cat__salary_listed_yes     0.228010
20009               cat__employment_type_part_time     0.184773
20019  cat__required_experience_director_executive     0.178285
20020               cat__required_experience_entry     0.076244
20007               cat__employment_type_full_time     0.003719
20002              cat__has_screening_questions_no    -0.039084
20016                cat__required_education_other    -0.073644
20011                 cat__employment_ty

In [26]:
import re
import pandas as pd
import sys
sys.path.append("../src")
from preprocessing import load_and_clean, combine_text_fields

df = combine_text_fields(load_and_clean("../data/emscad_core.csv"))

df["has_currency"] = df["full_text"].str.contains(r"[$₹€£]", regex=True)
df["has_url"] = df["full_text"].str.contains(r"http\S+|www\.\S+", regex=True)
df["urgency_count"] = df["full_text"].str.count(r"\b(urgent|immediately|apply now|limited spots|act now|hurry)\b", flags=re.IGNORECASE)
df["has_informal_contact"] = df["full_text"].str.contains(r"\b(whatsapp|telegram|text us|text me)\b", case=False, regex=True)
df["exclaim_count"] = df["full_text"].str.count("!")

for col in ["has_currency", "has_url", "has_informal_contact"]:
    print(f"\n--- {col} ---")
    print(df.groupby(col)["fraudulent"].agg(["mean", "count"]).round(3))

print("\n--- urgency_count (binned) ---")
print(df.groupby(pd.cut(df["urgency_count"], [-1, 0, 1, 2, 100]))["fraudulent"].agg(["mean", "count"]).round(3))

print("\n--- exclaim_count (binned) ---")
print(df.groupby(pd.cut(df["exclaim_count"], [-1, 0, 2, 5, 1000]))["fraudulent"].agg(["mean", "count"]).round(3))

C:\Users\pc\AppData\Local\Temp\ipykernel_31340\3555268493.py:12: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  df["has_informal_contact"] = df["full_text"].str.contains(r"\b(whatsapp|telegram|text us|text me)\b", case=False, regex=True)



--- has_currency ---
               mean  count
has_currency              
False         0.036  12560
True          0.080   2242

--- has_url ---
          mean  count
has_url              
False    0.043  14802

--- has_informal_contact ---
                       mean  count
has_informal_contact              
False                 0.043  14750
True                  0.000     52

--- urgency_count (binned) ---
                mean  count
urgency_count              
(-1, 0]        0.042  13614
(0, 1]         0.039   1117
(1, 2]         0.119     67
(2, 100]       0.500      4

--- exclaim_count (binned) ---
                mean  count
exclaim_count              
(-1, 0]        0.045  10110
(0, 2]         0.044   3556
(2, 5]         0.020    871
(5, 1000]      0.015    265
